### DISCUSSION TITLE
convert xarray Dataset to rioxarray Dataset?

### DISCUSSION THREAD

**Discussion by wendellwt** [corteva/rioxarray#691] [Q&A] [answered]
How can I convert xarray's xarray.DataSet to rioxarray's xarray.DataSet?

What I have works because I write out the xarray.DataSet (using xarray) and then read it back in (using rioxarray).  Both items say thay are an `<xarray.Dataset>`. Is there a better way than writing to a file & reading back in just to make it the rioxarray version of Dataset?

This works fine:

    ds_xa = xarray.open_dataset( infile,
                    decode_coords='all', engine='h5netcdf')

    # need xarray here because it has .coarsen()
    ds6_xa = ds_xa \
        .coarsen(x0=6, y0=6,boundary='trim') \
          .max()

    ds6_xa.to_netcdf('ds6.nc')  # q: why is this write needed?

    xr = rioxarray.open_rasterio('ds6.nc')  # q: can I do this directly to dx6_xa?

    vr = xr.VIL

    # need rioxarray here because it has .shapes()
    shapes = features.shapes(vr)

According to the [docs](https://corteva.github.io/rioxarray/stable/getting_started/getting_started.html), `.open_rasterio()` does some very nice things when reading a file from disk.

Q: how do I get rioxarray to do those things on an existing (i.e. in-memory) xarray DataSet?

**Comment by snowman2** [comment]
Is this what you are referring to: https://corteva.github.io/rioxarray/stable/getting_started/crs_management.html ?

**Comment by wendellwt** [comment]
Thank you, but no, it's not the crs part.  It's not as bad as I thought, and it may be in the "it's a feature, not a bug" category.

If I write out a Dataset that has this:
```
Dimensions:        (x: 6688, y: 3175)
Coordinates:
    grid_mapping0  int32 ...
  * x              (x) float64 -134.4 -134.3 -134.3 ... -61.66 -61.65 -61.64
  * y              (y) float64 53.87 53.86 53.85 53.84 ... 19.38 19.37 19.36
Data variables:
    start_time     datetime64[ns] ...
    stop_time      datetime64[ns] ...
    xml_metadata   |S47067 ...
    VIL            (y, x) float32 nan nan nan nan nan ... nan nan nan nan nan
    VIL_FLAGS      (y, x) float32 nan nan nan nan nan ... nan nan nan nan nan
    PRECIP_PHASE   (y, x) float32 nan nan nan nan nan ... nan nan nan nan nan
```
and read it back in, I get a Dataset that has this:
```
Dimensions:        (band: 1, x: 6688, y: 3175)
Coordinates:
  * band           (band) int64 1
  * x              (x) float64 -134.4 -134.3 -134.3 ... -61.66 -61.65 -61.64
  * y              (y) float64 53.87 53.86 53.85 53.84 ... 19.38 19.37 19.36
    grid_mapping0  int64 0
Data variables:
    VIL            (band, y, x) int16 ...
    VIL_FLAGS      (band, y, x) int8 ...
    PRECIP_PHASE   (band, y, x) int8 ...
```
It appears that:
* Band 1 was created
* data elements were changed from float32 to int8 and int16
* start_time, stop_time, and xml_metadata disappeared (I don't need these, btw)
* nan was changed to 0 (not sure about this one)

So, to replicate the write/read operation, it looks like `.VIL.fillna(0)` on the original Dataset would suffice.  Does this seem right?

**Comment by snowman2** [comment, marked as answer]
> data elements were changed from float32 to int8 and int16

This is because `mask_and_scale=True` is default with `open_dataset`. You can disable it if you don't want the data masked and scaled.

> nan was changed to 0 (not sure about this one)

This is because `mask_and_scale=True` is default with `open_dataset`. You can disable it if you don't want the data masked and scaled.

**Reply by wendellwt** [reply]
Thank you.  That's what I was looking for.  I just need to do those operations to the in-memory Dataset.
